Fix `remove_foreign_key` when multiple columns are found

zzak opened the pull request:
Fixes #55078

This PR has been updated to fix the behavior using 8.1 migration for backwards compatibilty.

/cc @kei-s

zzak committed:
Fix `remove_foreign_key` when multiple columns are found

Co-authored-by: fatkodima <fatkodima123@gmail.com>

fatkodima commented:
I wonder if we should make older migrations backwards compatible and how likely is the following case, because previously this worked (tables are from the issue)
```ruby
remove_foreign_key :books, :users # removes fkey on author_id
remove_foreign_key :books, :users, column: :user_id
```

and now will not work
```ruby
remove_foreign_key :books, :users # removes fkey on user_id
remove_foreign_key :books, :users, column: :user_id # raises because there is no foreign key
```

zzak commented:
@fatkodima Thanks for taking a look!

> remove_foreign_key :books, :users # removes fkey on author_id

I feel this is a bug and it should raise, but also that is kind of a breaking change.
My idea for `Compatibility` is that it's there for API changes to migration methods, or handling new options in a backwards compatible way, does that align with your understanding?

fatkodima commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1815,7 +1815,23 @@ def foreign_key_name(table_name, options)
           return unless use_foreign_keys?
-          foreign_keys(from_table).detect { |fk| fk.defined_for?(**options) }
+
+          if options[:_skip_column_match]
+            return foreign_keys(from_table).detect { |fk| fk.defined_for?(**options) }
+          end
+
+          keys = foreign_keys(from_table)
+
+          if options[:column]
+            expected_columns = Array(options[:column]).map(&:to_s)
+            keys.select! { |fk| Array(fk.options[:column]).map(&:to_s) == expected_columns }
+          elsif options[:to_table] && !options.key?(:primary_key) && keys.size > 1
+            default_column = foreign_key_column_for(options[:to_table], "id").to_s
+            match = keys.find { |fk| Array(fk.options[:column]).map(&:to_s) == [default_column] }
+            keys = [match].compact if match
+          end
+
+          keys.detect { |fk| fk.defined_for?(**options) }
```
Haven't tested, but something like this is shorter and should work too
```suggestion
          keys = foreign_keys(from_table)
          
          if keys.size == 1 || options[:_skip_column_match]
            keys.find { |fk| fk.defined_for?(**options) }
          elsif options[:column]
            expected_columns = Array(options[:column]).map(&:to_s)
            keys = keys.select { |fk| Array(fk.column) == expected_columns }
            keys.find { |fk| fk.defined_for?(**options) }
          else
            default_column = foreign_key_column_for(options[:to_table], "id")
            keys.find { |fk| fk.column == default_column }
          end
```

zzak replied:
That works, thanks for tidying it up!
Fixed in 02bcd2e

zzak commented:
@fatkodima I think this is ready for another review. Couple changes:

* Reverted the change to sqlite3, test coverage was a concern
* Added TableDefinition with tests (for change_table,etc apis)
* Left backwards compat path at the top to separate new logic
* Simplified branching instead of checking size for composite keys

We can try to circle back and fix sqlite, but I didn't want to hold this up anymore.

Thanks!

fatkodima reviewed (commented):
I didn't get why you decided to not fix sqlite in this PR? I think it will require only a few additional test cases and the PR will be of the same size as currently?

If you don't have time etc, let me know and I will try to implement it in this PR.

byroot committed:
Cleanup `compatible_table_definition` methods

No need to use `class <<`.
